### PR TITLE
Bring feature order and rendering order together in a useful way

### LINKED ISSUE #42428: 'Control feature rendering order' should have precedence over symbol levels
**Describe the bug**
'Control feature rendering order' should have precedence over symbol levels.
Now it isn't possible to create a map with for example correct road outlines and have roads in the correct order (bridge over road etc.).
For a detailed description and examples, see: https://gis.stackexchange.com/questions/204034/can-i-change-the-precedence-when-combining-control-feature-rendering-order-wit

**How to Reproduce**

1. Download OpenStreetMap sample data (for example shapefile from http://download.geofabrik.de/)
2. Open road-layer (shapefile) or open line-layer (Postgis / Osm2pgsql)
3. Add a rule-based style with an outline
4. Set symbol levels (0 and 1)
5. -> Now the map looks correctly (outlines are merged), except bridges over roads (the z-order isn't taken into account)
6. Check 'Control feature rendering order' and order ascending by 'layer' or 'z_order'
7. -> The map doesn't change
8. Delete symbol levels (set all to 0)
9. -> Map shows bridges correctly, but outlines are not merged at for example intersections.

**QGIS and OS versions**
QGIS version | 3.18.1-Zürich | QGIS code branch | Release 3.18
-- | -- | -- | --
Compiled against Qt | 5.15.2 | Running against Qt | 5.15.2
Compiled against GDAL/OGR | 3.0.4 | Running against GDAL/OGR | 3.0.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.35.2 | Running against SQLite | 3.35.2
PostgreSQL Client Version | 13.2 | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.6 | QScintilla2 Version | 2.12.1
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Arch Linux

**Additional context**
Please let me know if there are other possibilities to achieve the desired effect that I don't know.

### PR DESCRIPTION
**Pull request by joto**
There are two ways of taking "order" into account when rendering:

1. You can control the feature rendering order which essentially does an "order by" of your layer data.
2. You can enable symbol levels which defines which styles are rendered in which order.

This commit addresses a problem when trying to bring (1) and (2) together. When both are enabled, the rendering order (2) trumps the order given by (1), essentially ignoring the order from (1).

With this commit rendering instead starts "from scratch" every time the "order by" attributes have a new value.

This allows, for instance, to render complex roads networks correctly. Ordering via (1) is used to draw roads going over bridges etc. in the correct order while ordering via (2) is used to render road styles correctly (thicker dark line, then thinner light line to get a roads with casing).

The patch works by putting all features that need to be rendered not in a single QHash (Symbol->Feature) but a QList of those hashes. Every time at least one of the attributes used for the "order by" changes, a new element is added to that list. The patch is rather simple, it just looks a bit larger due to the extra indentation needed.

I thought about making this configurable, but I don't see any use case where it makes sense to have (1) and (2) enabled where you don't want this behaviour.

Fixes #42428
See #17276

### CODE REVIEW

**Comment by joto** [comment]
road-layering

Here you can see the change in action. The layering is all correct and only one QGIS layer was used.

**Comment by m-kuhn** [comment]
Thank you for opening this pull request.
The idea / logic sounds good to me and this use case will help produce nice cartographic results.

Could you do some syntactic changes to the code:

 - use `camelCase` variables
 - add a comment that explains the semantics of the new `QList` that was added (i.e. replace the term "array" with an explanation of what this list is used for)
 - add a comment that explains the meaning of the empty items in the list

These comments  will help the review process as well as for future maintenance.

It would also be nice to have a simple rendering test that covers this scenario. (see e.g. here: https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgs25drenderer.cpp).

File: src/core/vector/qgsvectorlayerrenderer.cpp [hunk 1]
```diff
@@ -542,7 +542,28 @@ void QgsVectorLayerRenderer::drawRendererLevels( QgsFeatureRenderer *renderer, Q
...
+  // Ordering is based on (a) a "level" which is determined by the configured
+  // feature rendering order" and (b) the symbol level. The "level" is
+  // determined by the values of the attributes defined in the feature
+  // rendering order settings. Each time the attribute(s) have a new distinct
+  // value, a new empty QHash is added to the "features" list. This QHash is
+  // then filled by mappings from the symbol to a list of all the features
+  // that should be rendered by that symbol.
+  //
+  // If orderBy is not enabled, this list will only ever contain a single
+  // element.
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayerrenderer.cpp:555]: thanks :heart:

**Comment by joto** [comment]
Thanks for your feedback. I have fixed the variable names and added some comments to the code. I need some more time to look at the tests and see what I need to do.

One potential problem I noticed while looking at the code again: If the attribute we do the *order by* on has many values, maybe even a different value for each feature, we'll generate an aweful lot of `QList` entries with one-element `QHash`es in them. This can use up a lot of memory. I don't think there is a way around that, and a user shouldn't do that, but if they do, performance will suffer. So maybe we have to detect that case in some way and handle it specially.

**Comment by joto** [comment]
@m-kuhn Any chance to get this reviewed?

**Comment by m-kuhn** [comment]
> One potential problem I noticed while looking at the code again: If the attribute we do the _order by_ on has many values, maybe even a different value for each feature, we'll generate an aweful lot of `QList` entries with one-element `QHash`es in them. This can use up a lot of memory. I don't think there is a way around that, and a user shouldn't do that, but if they do, performance will suffer. So maybe we have to detect that case in some way and handle it specially.

This is a fair point. Could you conduct a test with 10000 features to check the effect?

Another thing is the usage of std containers, compared to Qt containers which are used in the majority of cases. I am not sure if there's a guideline for that, so I'll leave it to others to speak up if there's an issue with that.

Apart from this, it looks good to me.

**Comment by joto** [comment]
> This is a fair point. Could you conduct a test with 10000 features to check the effect?

I tested it with a layer of >500,000 features using the unique id of the feature as "layer" (so it starts a new "layer" for each feature making this the worst case). It used more memory, but not that much and rendering speed wasn't much worse either, so I think we are good.

> Another thing is the usage of std containers, compared to Qt containers which are used in the majority of cases. I am not sure if there's a guideline for that, so I'll leave it to others to speak up if there's an issue with that.

I switched from `std::vector` to `QVector`.

**Review by m-kuhn — APPROVED** [review]
Thank you

**Comment by JohnProv** [comment]
Does this patch also concern rule-based rendering?
In 3.32.0 it works correctly with the categorized renderer, but not with the rule-based renderer. There the problem from #42428 still exists.

**Comment by joto** [comment]
@JohnProv I just checked this and you are right, it works in the categorized renderer, but not with the rule-based renderer. As far as I can see there is nothing specific to the renderer type in the code I changed. But I know next to nothing about the QGIS internals, just beat at the code until it worked for me. Clearly I missed something there. I suggest you reopen #42428 and put a comment there that it isn't fixed after all.

**Pull request merged**